include/onnxruntime/core/providers/dml/dml_provider_factory.h

```diff
@@ -44,6 +44,52 @@ ORT_API_STATUS(OrtSessionOptionsAppendExecutionProvider_DML, _In_ OrtSessionOpti
 ORT_API_STATUS(OrtSessionOptionsAppendExecutionProviderEx_DML, _In_ OrtSessionOptions* options,
                _In_ IDMLDevice* dml_device, _In_ ID3D12CommandQueue* cmd_queue);
 
+
+struct OrtDmlApi;
+typedef struct OrtDmlApi OrtDmlApi;
+
+const OrtDmlApi* GetOrtDmlApi(_In_ uint32_t version) NO_EXCEPTION;
+
+struct OrtDmlApi {
+  /**
+   * Creates a DirectML Execution Provider which executes on the hardware adapter with the given device_id, also known as
+   * the adapter index. The device ID corresponds to the enumeration order of hardware adapters as given by 
+   * IDXGIFactory::EnumAdapters. A device_id of 0 always corresponds to the default adapter, which is typically the 
+   * primary display GPU installed on the system. A negative device_id is invalid.
+  */
+  ORT_API2_STATUS(OrtSessionOptionsAppendExecutionProvider_DML, _In_ OrtSessionOptions* options, int device_id);
+
+  /**
+   * Creates a DirectML Execution Provider using the given DirectML device, and which executes work on the supplied D3D12
+   * command queue. The DirectML device and D3D12 command queue must have the same parent ID3D12Device, or an error will
+   * be returned. The D3D12 command queue must be of type DIRECT or COMPUTE (see D3D12_COMMAND_LIST_TYPE). If this 
+   * function succeeds, the inference session maintains a strong reference on both the dml_device and the command_queue 
+   * objects.
+   * See also: DMLCreateDevice
+   * See also: ID3D12Device::CreateCommandQueue
+   */
+  ORT_API2_STATUS(OrtSessionOptionsAppendExecutionProviderEx_DML, _In_ OrtSessionOptions* options,
+                _In_ IDMLDevice* dml_device, _In_ ID3D12CommandQueue* cmd_queue);
```

Comment:
>_In_ IDMLDevice* dml_device, _In_ ID3D12CommandQueue* cmd_queue [](http://example.com/codeflow?start=16&length=63)

Should we create a `OrtDmlProviderOptions` like all the other providers?

```
  ORT_API2_STATUS(SessionOptionsAppendExecutionProvider_CUDA,
                  _In_ OrtSessionOptions* options, _In_ const OrtCUDAProviderOptions* cuda_options);

  /** \brief Append ROCM execution provider to the session options
  *
  * If ROCM is not available (due to a non ROCM enabled build, or if ROCM is not installed on the system), this function will return failure.
  *
  * \param[in] options
  * \param[in] rocm_options
  *
  * \snippet{doc} snippets.dox OrtStatus Return Value
  */
  ORT_API2_STATUS(SessionOptionsAppendExecutionProvider_ROCM,
                  _In_ OrtSessionOptions* options, _In_ const OrtROCMProviderOptions* rocm_options);

  /** \brief Append OpenVINO execution provider to the session options
  *
  * If OpenVINO is not available (due to a non OpenVINO enabled build, or if OpenVINO is not installed on the system), this function will fail.
  *
  * \param[in] options
  * \param[in] provider_options
  *
  * \snippet{doc} snippets.dox OrtStatus Return Value
  */
  ORT_API2_STATUS(SessionOptionsAppendExecutionProvider_OpenVINO,
                  _In_ OrtSessionOptions* options, _In_ const OrtOpenVINOProviderOptions* provider_options);
```

Reply:
Probably, but this is likely a change for another PR?

Reply:
When you do create a struct just make sure it's versioned. We've not done a good job with this for other providers and hence recently changed this setup from defining the struct in the header to exposing an API that returns an opaque ptr to the struct and additional APIs to set values in the struct.

Reply:
I'm fine with not having a struct for now, but I don't like the silly "Ex" prefix pattern, because what will we call the next ones, ExEx and ExExEx? I think the Win32 standard guidance of choosing numbered functions over Ex functions is better here (e.g. `IDWriteRenderingParams3::GetRenderingMode1()`):

```
OrtSessionOptionsAppendExecutionProvider_DML // implicit 0
OrtSessionOptionsAppendExecutionProvider_DML1 // no Ex
OrtSessionOptionsAppendExecutionProvider_DML2 // no ExEx
OrtSessionOptionsAppendExecutionProvider_DML3 // no ExExEx :b
```

The typical guidance for struct versioning in Win32 C++ (e.g. DWrite's DWRITE_LINE_METRICS1) would be to add a new struct with new fields at the end, and overloads call the respective version of the struct, but for C, it's a bit more annoying because you have to copy the entire struct again (no struct inheritance) and name the function uniquely (no overloads). When we have more fields to initialize, the struct will become more useful.

Reply:
Now it is:
OrtSessionOptionsAppendExecutionProvider_DML
OrtSessionOptionsAppendExecutionProvider_DML1